src/charm.py

```diff
@@ -0,0 +1,323 @@
+import logging
+import os
+
+import ops.lib
+from charms.nginx_ingress_integrator.v0.nginx_route import require_nginx_route
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import ActiveStatus, WaitingStatus
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+pgsql = ops.lib.use("pgsql", 1, "postgresql-charmers@lists.launchpad.net")
+DATABASE_NAME = "synapse"
+
+VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
+
+
+class MatrixOperatorCharm(CharmBase):
+    """Charm the service."""
+
+    _CONTAINER_NAME = "synapse"
+    _SYNAPSE_CONFIG_PATH = "/data/homeserver.yaml"
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.start, self._on_install)
+        self.framework.observe(
+                self.on.synapse_pebble_ready,
+                self._on_config_changed
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.register_user_action, self._register_user)
+        require_nginx_route(
+            charm=self,
+            service_hostname=self._external_hostname,
+            service_name=self.app.name,
+            service_port=8008,
+        )
+        self._stored.set_default(
+            db_name=None,
+            db_user=None,
+            db_password=None,
+            db_host=None,
+            db_port=None,
+        )
+
+        self.db = pgsql.PostgreSQLClient(self, "db")  # pylint: disable=C0103
+        self.framework.observe(
+              self.db.on.database_relation_joined,
+              self._on_database_relation_joined
+          )
+        self.framework.observe(
+              self.db.on.master_changed,
+              self._on_master_changed
+          )
+
+    def _on_database_relation_joined(
+         self, event: pgsql.DatabaseRelationJoinedEvent  # type: ignore
+    ) -> None:
+        """Handle db-relation-joined.
+
+        Args:
+            event: Event triggering the database relation joined handler.
+        """
+        if self.model.unit.is_leader():
+            # Provide requirements to the PostgreSQL server.
+            event.database = DATABASE_NAME
+            event.extensions = ["hstore:public", "pg_trgm:public"]
+        elif event.database != DATABASE_NAME:
+            # Leader has not yet set requirements. Defer, in case this unit
+            # becomes leader and needs to perform that operation.
+            event.defer()
+
+    # pgsql.MasterChangedEvent is actually defined
+    def _on_master_changed(self, event: pgsql.MasterChangedEvent) -> None:
+        """Handle changes in the primary database unit.
+
+        Args:
+            event: Event triggering the database master changed handler.
+        """
+        if event.master is None:
```

Comment:
The `woke` tool probably won't like this

Reply:
It actually won't like `pgsql.MasterChangedEvent` either

Reply:
It won't like at least two Metallica albums either...